File: proxyshield/src/utils/logger.js
let verbose = false;

/**
 * Set verbose mode. When enabled, debug() calls produce output.
 * @param {boolean} enabled - Whether to enable verbose/debug logging
 */
export function setVerbose(enabled) {
  verbose = enabled;
}

/**
 * Write a structured JSON log entry to the given output stream.
 * @param {NodeJS.WriteStream} stream - process.stdout or process.stderr
 * @param {string} level - Log level (info, warn, error, debug)
 * @param {string} message - Log message
 * @param {Object} [data] - Optional additional fields to include in the log entry
 */
function write(stream, level, message, data) {
  const entry = {
    level,
    message,
    timestamp: new Date().toISOString(),
    ...data
  };
  stream.write(JSON.stringify(entry) + '\n');
}

/**
 * Structured JSON logger for ProxyShield.
 * All internal logging goes through this — never use console.log in library code.
 */
const logger = {
  /**
   * Log an informational message to stdout.
   * @param {string} message - The log message
   * @param {Object} [data] - Optional additional fields
   */
  info(message, data) {
    write(process.stdout, 'info', message, data);
  },

  /**
   * Log a warning message to stderr.
   * @param {string} message - The log message
   * @param {Object} [data] - Optional additional fields
   */
  warn(message, data) {
    write(process.stderr, 'warn', message, data);
  },

  /**
   * Log an error message to stderr.
   * @param {string} message - The log message
   * @param {Object} [data] - Optional additional fields
   */
  error(message, data) {
    write(process.stderr, 'error', message, data);
  },

  /**
   * Log a debug message to stdout. Only produces output when verbose mode is enabled.
   * @param {string} message - The log message
   * @param {Object} [data] - Optional additional fields
   */
  debug(message, data) {
    if (!verbose) return;
    write(process.stdout, 'debug', message, data);
  }
};

export default logger;

File: proxyshield/proxyshield-core/internal/logger/logger.go
// Package logger provides a structured JSON logger used throughout the proxy.
package logger

import (
	"encoding/json"
	"os"
	"sync"
	"time"
)

// Field represents a key-value pair for structured logging.
type Field struct {
	Key   string
	Value interface{}
}

// F creates a new Field with the given key and value.
func F(key string, value interface{}) Field {
	return Field{Key: key, Value: value}
}

// Logger writes structured JSON log entries.
type Logger struct {
	mu      sync.Mutex
	verbose bool
	silent  bool
}

func (l *Logger) write(w *os.File, level, msg string, fields []Field) {
	if l.silent {
		return
	}
	entry := make(map[string]interface{}, len(fields)+3)
	entry["level"] = level
	entry["message"] = msg
	entry["timestamp"] = time.Now().UTC().Format(time.RFC3339)
	for _, f := range fields {
		entry[f.Key] = f.Value
	}

	data, err := json.Marshal(entry)
	if err != nil {
		return
	}

	l.mu.Lock()
	defer l.mu.Unlock()
	w.Write(data)
	w.Write([]byte("\n"))
}

// Info logs an informational message to stdout.
func (l *Logger) Info(msg string, fields ...Field) {
	l.write(os.Stdout, "info", msg, fields)
}

// Warn logs a warning message to stderr.
func (l *Logger) Warn(msg string, fields ...Field) {
	l.write(os.Stderr, "warn", msg, fields)
}

// Error logs an error message to stderr.
func (l *Logger) Error(msg string, fields ...Field) {
	l.write(os.Stderr, "error", msg, fields)
}

// Debug logs a debug message to stdout only if verbose mode is enabled.
func (l *Logger) Debug(msg string, fields ...Field) {
	if l.verbose {
		l.write(os.Stdout, "debug", msg, fields)
	}
}

// defaultLogger is the package-level logger instance.
var defaultLogger = &Logger{}

// verbose controls debug output for the default logger.
var verbose bool

// SetVerbose enables or disables debug logging.
func SetVerbose(v bool) {
	verbose = v
	defaultLogger.verbose = v
}

// Silence suppresses all log output when set to true. Used in benchmark mode.
func Silence(s bool) {
	defaultLogger.silent = s
}

// Info logs an informational message using the default logger.
func Info(msg string, fields ...Field) {
	defaultLogger.Info(msg, fields...)
}

// Warn logs a warning message using the default logger.
func Warn(msg string, fields ...Field) {
	defaultLogger.Warn(msg, fields...)
}

// Error logs an error message using the default logger.
func Error(msg string, fields ...Field) {
	defaultLogger.Error(msg, fields...)
}

// Debug logs a debug message using the default logger.
func Debug(msg string, fields ...Field) {
	defaultLogger.Debug(msg, fields...)
}
